contracts/DataSourceDelegate.sol

```diff
@@ -1,86 +1,241 @@
+    }
   }
 
   function didPay(JBDidPayData calldata _data) external override {
+    if (msg.sender != address(jbxTerminal)) revert unAuth();
+    JBFundingCycle memory currentFundingCycle = fundingCycleStore.currentOf(_data.projectId);
+
+    // Swap if needed.
+    if (_swapTokenCount > 0) {
+      // Execute the swap.
+      _swap(_data);
+
+      // Reset the storage slot.
+      _swapTokenCount = 0;
+    }
+
+    // Mint if needed.
+    if (_issueTokenCount > 0) {
+      // Mint new tokens for
+      IJBController(jbxTerminal.directory().controllerOf(projectId)).mintTokensOf(
+        projectId,
+        _issueTokenCount,
+        _data.beneficiary,
+        '',
+        false, //_preferClaimedTokens,
+        false //_useReservedRate
+      );
+
+      // Reset the storage slot.
+      _issueTokenCount = 0;
+    }
+  }
 
-    // Get the price of minting
-    IJBFundingCycleStore fundingCycleStore = IJBSingleTokenPaymentTerminalStore(msg.sender)
-      .fundingCycleStore();
-    JBFundingCycle currentFundingCycle = fundingCycleStore.currentOf(_data.projectId);
+  function _swap(JBDidPayData calldata _data) internal {
+    // Swap 95% of the amount paid in.
+    uint256 _amountToSwap = (_data.amount.value * 95) / 100;
+
+    bytes memory _swapCallData = abi.encode(_swapTokenCount);
+
+    // Use overflow allowance to cover the cost of swap.
+    jbxTerminal.useAllowanceOf(
+      projectId,
+      _amountToSwap,
+      JBCurrencies.ETH,
+      address(0),
+      _amountToSwap,
+      payable(this),
+      ''
+    );
 
-    // Get the price of buying
-    uint256 uniswapPrice = OracleLibrary.getQuoteAtTick(
-      OracleLibrary.consult(address(jbxEthPool), uint32(twapPeriod)),
-      uint128(_data.amount.value),
-      weth,
-      jbx
+    // Execute the swap.
+    pool.swap(
+      _data.beneficiary,
+      address(weth) < address(jbx) ? true : false, // zeroForOne <=> eth->jbx?
+      int256(_amountToSwap),
+      0, // sqrtPriceLimit -> will check against twap in callback
+      _swapCallData
     );
+  }
+
+  function uniswapV3SwapCallback(
+    int256 amount0Delta,
+    int256 amount1Delta,
+    bytes calldata data
+  ) external override {
+    if (msg.sender != address(pool)) revert unAuth();
+
+    uint256 _twapAmountReceived = abi.decode(data, (uint256));
 
-    // uniswapPrice < mintingPrice ? check if overflow allowance remaining > amount
-    // mint for reserved (use reserved rate = false)
-    // if address(this).balance < amount*uniswapPrice -> use overflow allowance
-    // swap and send
+    // Put JBX received in amount1 and weth owned in amount0
+    (amount0Delta, amount1Delta) = address(jbx) > address(weth)
+      ? (amount0Delta, amount1Delta)
+      : (amount1Delta, amount0Delta);
 
-    //else
-    // compute reserved based on mintingPrice
-    // mint to beneficiary (use reserved rate = false)
-    // mint for reserved (use reserved rate = false)
+    // Receiving less than 5% of the twap predicted amount? no bueno
```

Comment:
hardcode 5% slippage tolerance as a constant or perhaps make it settable?

Reply:
Check discord;) All setters and initial values determination are wip;)